### ISSUE TITLE
Schema is not applied in reading 

### ISSUE THREAD

**Issue by emadb** [Automattic/mongoose#4279]
I have a collection on MongoDb with several fields and not all of them are defined in the model schema. For example on the db I have a document with fields `a, b, c` and `d`, and in  my mongoose schema I have only `a` and `b` defined. 
When I query the collection, the returned documents contains also these additional fields. 
Why? IMHO it shouldn't be like that, the returned document should contains only the declared fields.
Opinions?

thx

**Comment by vkarpov15** [comment]
By design, mongoose only validates things going into the db, it assumes everything coming out is already valid

**Comment by emadb** [comment]
Ok, I understand. Do you think it would be hard to load (or at least serialize) only the fields defined in the schema? 
Sometimes the web app don't own the database (other apps are using it) and it became hard to override the toJson method on every model to remove the unwanted fields.

Every ORM that I used in the past work like that.

**Comment by vkarpov15** [comment]
Fair point. Perhaps we can add an option to filter out fields that aren't in the schema on the way out of the DB in a future release. You could probably implement a plugin to do this using a pre init hook...

**Comment by emadb** [comment]
For my special case I wrote this plugin:
```
function cleaner(schema) {
  schema.post('init', function(doc) {
    const schemaKeys = Object.keys(schema.paths).map(p => p.split('.')[0]) // only first level attrs
    Object.keys(doc._doc).forEach(f => {
      if (schemaKeys.indexOf(f) === -1){
        delete doc._doc[f];  
      }
    })
  })
}
```

I don't like to access to _doc internal object but I didn't find a way to avoid it.
Any feedback is welcome.

**Comment by vkarpov15** [comment]
@MidnightP

**Comment by MidnightP** [comment]
thanks! 

@emadb  Have you tried ```Model.findOne().then(doc => new Model(doc))```?

**Comment by MidnightP** [comment]
@vkarpov15 

Considering the announced possibility of ["Schema validation with JSON Schema for new data governance controls"](https://www.mongodb.com/webinar/whats-new-in-mongodb-3-6-emea), Will there be any changes in the way Mongoose was designed with the release of Mongo 3.6?

**Comment by vkarpov15** [comment]
@MidnightP no plans to integrate with that currently. We considered using the [document validation functionality that was introduced in 3.2](https://docs.mongodb.com/manual/core/document-validation/), but that was too much of a stretch to put into core mongoose because then we'd have to have some mechanism for migrations and for managing cases where one application has one version of the schema and another application has another version.

The new JSON schema stuff coming in 3.6 is cool but I don't see it playing well with mongoose for similar reasons. The most I could see us doing right now is writing a plugin to help you convert mongoose schemas into something the mongodb server understands.

**Comment by libook** [comment]
@vkarpov15 
JSON Schema can be use to validate JSON data in much places in modern web services.  It will be very helpful if there are tools for converting format between Mongoose Schemas and JSON Schemas. That will be easy to reuse JSON Schemas.

**Comment by Enado95** [comment]
> @MidnightP no plans to integrate with that currently. We considered using the [document validation functionality that was introduced in 3.2](https://docs.mongodb.com/manual/core/document-validation/), but that was too much of a stretch to put into core mongoose because then we'd have to have some mechanism for migrations and for managing cases where one application has one version of the schema and another application has another version.
> 
> The new JSON schema stuff coming in 3.6 is cool but I don't see it playing well with mongoose for similar reasons. The most I could see us doing right now is writing a plugin to help you convert mongoose schemas into something the mongodb server understands.

I see this is still an open issue. Is there plans to implement a plugin to read only fields defined in the schema?

**Comment by vkarpov15** [comment]
@Enado95 not currently.

**Comment by som14062005** [comment]
Hi @vkarpov15, I'd like to implement this feature as discussed. Is this still something you'd accept as a PR? I'm planning to add a schema option (e.g., strictQuery: 'filter') that filters unknown fields during document init. Would this approach work?

**Comment by vkarpov15** [comment]
@som14062005 No that approach wouldn't work well - `strictQuery` already has other semantics. What if you use `strictQuery: 'throw'` but also want the `strictQuery: 'filter'` behavior? This would need to be a separate option at minimum.

**Comment by som14062005** [comment]
@vkarpov15 Thanks for the clarification!

Agreed — overloading `strictQuery` would create conflicting semantics.  
It seems cleaner to introduce a **separate schema option** specifically for the document read/init path.

For example, something like `strictRead` (or `strictInit`) that applies when hydrating documents from MongoDB:

```js
const schema = new Schema(
  {
    name: String,
    email: String
  },
  {
    strictRead: 'filter' // proposal: 'filter', 'throw', or false (default)
  }
);

**Comment by vkarpov15** [comment]
Yeah `strictRead` sounds like it could work
